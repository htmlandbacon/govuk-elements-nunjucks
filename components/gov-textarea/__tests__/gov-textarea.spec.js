'use strict';

const nunjucks = require('nunjucks');
const cheerio = require('cheerio');

describe('gov-textarea', () => {
  it('should set the ID as textarea-{name} if ID attr not supplied', () => {
    const name = 'more-details';
    const output = nunjucks.render('./components/gov-textarea/template.njk', {
      label: 'More details',
      name
    });
    const $ = cheerio.load(output);
    const textareaId = $('textarea').attr('id');

    expect(textareaId).toBe(`textarea-${name}`);
    expect(output).toMatchSnapshot();
  });

  it('should use ID attribute value over generated textarea-{name} ID', () => {
    const id = 'my-id';
    const output = nunjucks.render('./components/gov-textarea/template.njk', {
      label: 'More details',
      name: 'more-details',
      id
    });
    const $ = cheerio.load(output);
    const textareaId = $('textarea').attr('id');

    expect(textareaId).toBe(id);
    expect(output).toMatchSnapshot();
  });

  it('should add a form hint using the hint attribute', () => {
    const hint = 'Do not include bank details';
    const output = nunjucks.render('./components/gov-textarea/template.njk', {
      label: 'More details',
      name: 'more-details',
      hint
    });
    const $ = cheerio.load(output);
    const formHintText = $('label .form-label-bold + span.form-hint').html();

    expect(formHintText).toBe(hint);
    expect(output).toMatchSnapshot();
  });

  it('should add maxlength to textarea using the maxlength attribute', () => {
    const maxlength = '1200';
    const output = nunjucks.render('./components/gov-textarea/template.njk', {
      label: 'More details',
      name: 'more-details',
      maxlength
    });
    const $ = cheerio.load(output);
    const textareaMaxlength = $('textarea').attr('maxlength');

    expect(textareaMaxlength).toBe(maxlength);
    expect(output).toMatchSnapshot();
  });

  it('should add rows attribute to textarea using the rows attribute', () => {
    const rows = '12';
    const output = nunjucks.render('./components/gov-textarea/template.njk', {
      label: 'More details',
      name: 'more-details',
      rows
    });
    const $ = cheerio.load(output);
    const textareaRows = $('textarea').attr('rows');

    expect(textareaRows).toBe(rows);
    expect(output).toMatchSnapshot();
  });

  it('should set the value of the textarea using the value attribute', () => {
    const value = 'I really loved this service, it was great';
    const output = nunjucks.render('./components/gov-textarea/template.njk', {
      label: 'More details',
      name: 'more-details',
      value
    });
    const $ = cheerio.load(output);
    const textareaText = $('textarea').text();

    expect(textareaText).toBe(value);
    expect(output).toMatchSnapshot();
  });

  it('should add error message and classes when passed an error object', () => {
    const error = 'Put something in the box';
    const name = 'more-details';
    const output = nunjucks.render('./components/gov-textarea/template.njk', {
      label: 'More details',
      name,
      error
    });
    const $ = cheerio.load(output);
    const formGroupClasses = $('.form-group').attr('class');
    const errorMsg = $('label .form-label-bold + span.error-message').text();

    expect(formGroupClasses).toBe('form-group form-group-error');
    expect(errorMsg).toBe(error);
    expect(output).toMatchSnapshot();
  });
  describe('class options', () => {
    it('should add classes to textarea/label when supplied', () => {
      const classes = {label: 'test-label-class', input: 'test-input-class'};
      const output = nunjucks.render('./components/gov-textarea/template.njk', {
        label: 'Full name',
        classes
      });
      const $ = cheerio.load(output);
      const labelClasses = $('.form-group label span').attr('class');
      const inputClasses = $('.form-group textarea').attr('class');

      expect(labelClasses).toBe('test-label-class');
      expect(inputClasses).toBe('form-control test-input-class');
      expect(output).toMatchSnapshot();
    });
    it('should default to class form-label-bold when label class is not supplied', () => {
      const classes = {input: 'test-input-class'};
      const output = nunjucks.render('./components/gov-textarea/template.njk', {
        label: 'Full name',
        classes
      });
      const $ = cheerio.load(output);
      const labelClasses = $('.form-group label span').attr('class');
      const inputClasses = $('.form-group textarea').attr('class');

      expect(labelClasses).toBe('form-label-bold');
      expect(inputClasses).toBe('form-control test-input-class');
      expect(output).toMatchSnapshot();
    });
  });
});
