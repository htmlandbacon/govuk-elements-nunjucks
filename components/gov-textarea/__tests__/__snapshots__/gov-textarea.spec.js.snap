// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gov-textarea class options should add classes to textarea/label when supplied 1`] = `
<div id="textarea-undefined-form" class="form-group">
  <label for="textarea-undefined">
    <span class="test-label-class">Full name</span>
  </label>
  <textarea class="form-control test-input-class" id="textarea-undefined"
  rows="8" name=""></textarea>
</div>
`;

exports[`gov-textarea class options should default to class form-label-bold when label class is not supplied 1`] = `
<div id="textarea-undefined-form" class="form-group">
  <label for="textarea-undefined">
    <span class="form-label-bold">Full name</span>
  </label>
  <textarea class="form-control test-input-class" id="textarea-undefined"
  rows="8" name=""></textarea>
</div>
`;

exports[`gov-textarea should add a form hint using the hint attribute 1`] = `
<div id="textarea-more-details-form" class="form-group">
  <label for="textarea-more-details">
    <span class="form-label-bold">More details</span>
    <span class="form-hint">Do not include bank details</span>
  </label>
  <textarea class="form-control" id="textarea-more-details" rows="8" name="more-details"></textarea>
</div>
`;

exports[`gov-textarea should add error message and classes when passed an error object 1`] = `
<div id="textarea-more-details-form" class="form-group form-group-error">
  <label for="textarea-more-details">
    <span class="form-label-bold">More details</span>
    <span class="error-message">Put something in the box</span>
  </label>
  <textarea class="form-control" id="textarea-more-details" rows="8" name="more-details"></textarea>
</div>
`;

exports[`gov-textarea should add maxlength to textarea using the maxlength attribute 1`] = `
<div id="textarea-more-details-form" class="form-group">
  <label for="textarea-more-details">
    <span class="form-label-bold">More details</span>
  </label>
  <textarea class="form-control" id="textarea-more-details" rows="8" maxlength="1200"
  name="more-details"></textarea>
</div>
`;

exports[`gov-textarea should add rows attribute to textarea using the rows attribute 1`] = `
<div id="textarea-more-details-form" class="form-group">
  <label for="textarea-more-details">
    <span class="form-label-bold">More details</span>
  </label>
  <textarea class="form-control" id="textarea-more-details" rows="12" name="more-details"></textarea>
</div>
`;

exports[`gov-textarea should set the ID as textarea-{name} if ID attr not supplied 1`] = `
<div id="textarea-more-details-form" class="form-group">
  <label for="textarea-more-details">
    <span class="form-label-bold">More details</span>
  </label>
  <textarea class="form-control" id="textarea-more-details" rows="8" name="more-details"></textarea>
</div>
`;

exports[`gov-textarea should set the value of the textarea using the value attribute 1`] = `
<div id="textarea-more-details-form" class="form-group">
  <label for="textarea-more-details">
    <span class="form-label-bold">More details</span>
  </label>
  <textarea class="form-control" id="textarea-more-details" rows="8" name="more-details">I really loved this service, it was great</textarea>
</div>
`;

exports[`gov-textarea should use ID attribute value over generated textarea-{name} ID 1`] = `
<div id="my-id-form" class="form-group">
  <label for="my-id">
    <span class="form-label-bold">More details</span>
  </label>
  <textarea class="form-control" id="my-id" rows="8" name="more-details"></textarea>
</div>
`;
